Document GraphQL Handler configuration, logger options, and security measures (#743)
redwoodjs/redwoodjs.com: docs/graphql.md

- dt has some thoughts on this
- insomnia -->

## Logging

Logging is essential in production apps to be alerted about critical errors and to be able to respond effectively to support issues. In staging and development environments, logging helps you debug queries, resolvers and cell requests.

We want to make logging simple when using RedwoodJS and therefore have configured the api-side GraphQL handler to log common information about your queries and mutations. Log statements also be optionally enriched with [operation names](https://graphql.org/learn/queries/#operation-name), user agents, request ids, and performance timings to give you move visibility into your GraphQL api.

By configuring the GraphQL handler to use your api side [RedwoodJS logger](https://redwoodjs.com/docs/logger), any errors and other log statements about the [GraphQL execution](https://graphql.org/learn/execution/) will be logged to the [destination](https://redwoodjs.com/docs/logger#destination-aka-where-to-log) you've set up: to standard output, file, or transport stream.

> You configure the logger using the `loggerConfig` that accepts a [`logger`]((https://redwoodjs.com/docs/logger)) and s set of [GraphQL Logger Options](#graphql-logger-options):

Reviewer: Think 232-236 is repeated from the next section

Author: Sorry, I don't follow. Can you explain given the new updated document? Thanks!

Reviewer: I just mean line 232-236 is repeated content, because you're showing how to configure a logger in line 249.

Author: Got it. removing.

```js
  loggerConfig: { logger, options: {} }
```

### Configure the GraphQL Logger

> A typical GraphQLHandler `graphql.ts` is as follows:

Reviewer: This'll go out of date very quickly I think... maybe just have a smaller snippet passing loggerConfig with some options to createGraphQlHandler?

Author: Makes sense. I take that approach in a few places later in the doc -- just that option itself.

```js
// api/src/functions/graphql.ts

import {
  createGraphQLHandler,
  makeMergedSchema,
  makeServices,
} from '@redwoodjs/api'

import schemas from 'src/graphql/**/*.{js,ts}'
import { db } from 'src/lib/db'
import { logger } from 'src/lib/logger'
import services from 'src/services/**/*.{js,ts}'

export const handler = createGraphQLHandler({
  loggerConfig: { logger, options: {} },
  schema: makeMergedSchema({
    schemas,
    services: makeServices({ services }),
  }),
  onException: () => {
    // Disconnect from your database with an unhandled exception.
    db.$disconnect()
  },
})
```

#### Log Common Information

The `loggerConfig` takes several options that logs meaningful information along the graphQL execution lifecycle.

> ```js

Reviewer: JSDocs are great for highlights, but terrible for readibility in documentation imo (too verbose). Do you think we  can just add a table here with each of the options and describing them?

Author: Yes, a table is a much better idea. Thanks!

   */
  operationName?: boolean

  /**
   * @description Include the event's requestId, or if none, generate a uuid as an identifier.
   *
   * The requestId can be helpful when contacting your deployment provider to resolve issues when encountering errors or unexpected behavior.
   */
  requestId?: boolean

  /**
   * @description Include the query. This is the query or mutation (with fields) made in the request.
   */
  query?: boolean

  /**
   * @description Include the tracing and timing information.
   *
   * This will log various performance timings withing the GraphQL event lifecycle (parsing, validating, executing, etc).
   */
  tracing?: boolean

  /**
   * @description Include the browser (or client's) user agent.
   *
   * This can be helpful to know what type of client made the request to resolve issues when encountering errors or unexpected behavior.
   */
  userAgent?: boolean
}
```

Therefore, if you wish to log the GraphQL `query` made, the `data` returned, and the `operationName` used, you would

```js
// api/src/functions/graphql.ts

export const handler = createGraphQLHandler({
  loggerConfig: {
    logger,

>     options: { data: true, operationName: true, query: true },

Reviewer: Using a diff here and just showing the 3/4 relevant lines here would probably make it more digestable.

#### RequestId for Support Issue Resolution

Often times, your deployment provider will provide a request identifier to help reconcile and track down problems at an infrastructure level. For example, AWS API GAteway and AWS Lambda (used by Netlify, for example) provides `requestId` on the `event`.

You can include the request identifier setting the `requestId` logger option to `true`.

```js
// api/src/functions/graphql.ts
// ...
export const handler = createGraphQLHandler({
  loggerConfig: { logger, options: { requestId: true } },
// ...
```

And then, when working to resolve a support issue with your deployment provider, you can supply this request id to help them track down and investigate the problem more easily.
#### No Need to Log within Services

If you configure your GraphQL logger to include `data` and `query` information about each request adn its response as shown in:

```js
// api/src/functions/graphql.ts
// ...
export const handler = createGraphQLHandler({
  loggerConfig: { logger, options: { data: true, operationName: true, query: true } },
// ...
```

> then you do not need to log your `input` variables or the results in each service method.

Reviewer: For me this whole section on what you should not do doesn't need to be here :)

Author: Perhaps I can rephrase as a suggestion or recommendation.

I have seen on more than one occasion people in the Community take this approach to debugging services and query output:

```js
export const user = () => {
  if (context?.currentUser?.sub) {
    return db.user
      .findUnique({
        where: { id: context?.currentUser?.sub  }
      })
      .then(thenDebug('user'))

  } else {
    return Promise.resolve(null)
  }
};
```

and then

```js
const stringify = require('json-stringify-safe');

import { logger } from './logger'

export const thenDebug = (label) => (data) => {
  logger.debug(`${label}: ${stringify(data, null, 2)}`)
  return data
}
```

And I want to show that it's note necessary to have all that clutter in *every* service and still get the same debugging log info that helps.

That's my intention ... now how best to communicate that. Hm.

Reviewer: Yeah, the problem with including "dont do this" snippets is that its very easy to miss the "dont" part.

Just adding a sentence at the end to say this will log whatever you need, without needing to add additional logger statements into your services

Author: Got it. Yes, I can see how by showing example of what not to do, people might think of it as something to do (by accident).

I made some revisions.

api | INFO [2021-07-09 14:20:12.114 +0000] (apollo-graphql-server): GraphQL willSendResponse
api |     data: {
api |       "post": {
api |         "id": 2,
api |         "title": "Lime Tree Arbour",
api |         "body": "The wind in the trees is whispering \\ Whispering low that I love her \\ She puts her hand over mine \\ Down in the lime tree arbour",
api |         "createdAt": "2021-03-18T05:32:39.258Z",
api |         "publishedAt": "2021-03-18T05:32:39.258Z",
api |         "updatedAt": "2021-07-09T01:52:08.005Z",
api |         "__typename": "Post"
api |       }
api |     }
api |     query: "query ($id: Int!) {\n  post(id: $id) {\n    id\n    title\n    body\n    createdAt\n    publishedAt\n    updatedAt\n    __typename\n  }\n}\n"
```

but keep your services concise!

#### Send to Third-party Transports

Stream to third-party log and application monitoring services vital to production logging in serverless environments like [logFlare](https://logflare.app/), [Datadog](https://www.datadoghq.com/) or [LogDNA](https://www.logdna.com/)

#### Supports Log Redaction

> Everyone has herd or reports that Company X logged emails, or passwords to files or systems that may not have been secured. While RedwoodJS logging won't necessarily prevent that, it does provide you with the mechanism to ensure that won't happen.

Reviewer: herd -> heard

To redact sensitive information, you can supply paths to keys that hold sensitive data using the RedwoodJS logger [redact option](https://redwoodjs.com/docs/logger#redaction).

Because this logger is used with the GraphQL handler, it will respect any redaction paths setup.

For example, you have chosen to log `data` return by each request, then you may want to redact sensitive information, like email addresses from yur logs.

Here is an example of an application `/api/src/lib/logger.ts` confgured to redact email addresses. Take note of the path `data.users[*].email` as this says, in the `data` attribute, redact the `email` from every `user`:

```js
// /api/src/lib/logger.ts
import { createLogger } from '@redwoodjs/api/logger'

> import { redactionsList } from '@redwoodjs/api/logger'

Reviewer: Suggested snippet: 
```diff
// /api/src/lib/logger.ts
+import { createLogger, redactionsList } from '@redwoodjs/api/logger'

export const logger = createLogger({
  options: {
+    redact: [...redactionsList, 'email', 'data.users[*].email'],
  },
})
```

Author: Agreed. Thanks.

 *
 * These extend and override default LoggerOptions,
 * can define a destination like a file or other supported pin log transport stream,
 * and sets where or not to show the logger configuration settings (defaults to false)
 *
 * @param RedwoodLoggerOptions
 *
 * RedwoodLoggerOptions have
 * @param {options} LoggerOptions - defines how to log, such as pretty printing, redaction, and format
 * @param {string | DestinationStream} destination - defines where to log, such as a transport stream or file
 * @param {boolean} showConfig - whether to display logger configuration on initialization
 */
export const logger = createLogger({
  options: {
    prettyPrint: true,
    redact: [...redactionsList, 'email', 'data.users[*].email'],
  },
})
```

#### Timing Traces for Benchmarking Performances

Often you want to measure and report how long your queries take to execute and respond. You may already be measuring these durations at the database level, but you can also measure the time it takes for your the GraphQL server to parse, validate, and execute the request.

You may turn on logging these metrics via the `tracing` GraphQL configuration option.

> ```

Reviewer: Too much repeated info here I think. Just showing 
```
createGraphQlHandler({
	loggerOptions: {logger: {tracing: true}}
	/*
	..other options
	*/
```

Probably more helpful

Author: Agreed

  /**
   * @description Include the tracing and timing information.
   *
   * This will log various performance timings withing the GraphQL event lifecycle (parsing, validating, executing, etc).
   */
  tracing?: boolean
  ```

Let's say we wanted to get some benchmark numbers for the "find post by id" resolver

```js
  return await db.post.findUnique({
    where: { id },
  })
```

We see that this request took about 500 msecs (note: duration is reported in nanoseconds).

Ror more details about the information logged and its format, see [Apollo Tracing](https://github.com/apollographql/apollo-tracing).

```terminal
pi | INFO [2021-07-09 14:25:52.452 +0000] (apollo-graphql-server): GraphQL willSendResponse

> api |     tracing: {

Reviewer: Could we shotten this output just to include the important bits? Too much stuff in the documentation feels very daunting

Author: :+1: I'll "ellipses" out several of the resolver path timings ... the top level duration is the most important.

The [GraphQL Playground](https://github.com/graphql/graphql-playground) is a way for you to interact with your schema and try out queries and mutations. It can show you the schema by inspecting it. You can find the GraphQL Playground at http://localhost:8911/graphql when your dev server is running.

> Because both introspection and the playground share possibly sensitive information about your data model, your data, your queries and mutations, best practices for deploying a GraphQL Server call to disable these in production, RedwoodJS **only enables introspection and the playground when running in development**. That is when `process.env.NODE_ENV === 'development'`.

### Query Depth Limit

Attackers often submit expensive, nested queries that could overload your database or expend costly resources.

Typically, these types of  complex and expensive queries are usually huge deeply nested and take advantage of an understanding of your schema (hence why schema introspection is disabled byu default in production) and the data model relationships to create "cyclical" queries.

> Such unbounded GraphQL queries allow attackers to abuse query depth and with enough depth, this can easily impact your Graphql server and application.

Reviewer: ```suggestion
```

Author: Too scary? It's true though.

Reviewer: Nah I think you've mentioned this info in the above and below paragraphs

Author: Combined the first sentence and this one to use some same terms, but shorten this section.

An example of a cyclical query here takes advantage of knowing that and author has posts and each post has and author ... that has posts ... that has an another that ... etc.

This cyclical query has a depth of 8.

```js
// cyclical query example
// depth: 8+
query cyclical {
  author(id: 'jules-verne') {
    posts {
      author {
        posts {
          author {
            posts {
              author {
                ... {
                  ... # more deep nesting!
                }
              }
            }
          }
        }
      }
    }
  }
}
```

> To mitigate the risk of attacking your application via deeply nested queries, RedwoodJS by default sets the [Query Depth Limit](https://www.npmjs.com/package/graphql-depth-limit#documentation) to 11. 

> If You would like to set the limit to a lower or higher limit, you may do so via the `depthLimitOptions` setting when creating your GraphQL handler.

Reviewer: ```suggestion
You can change the default value via the `depthLimitOptions` setting when creating your GraphQL handler.
```

Author: Yup

> ```

Reviewer: We can take this jsdoc out I think

Author: Agreed

  /**
   * Limit the complexity of the queries solely by their depth.
   * @see https://www.npmjs.com/package/graphql-depth-limit#documentation
   */
  depthLimitOptions?: DepthLimitOptions
```

You `depthLimitOptions` are `maxDepth` or `ignore` stops recursive depth checking based on a field name. Ignore can be [either a string or regexp]( https://www.npmjs.com/package/graphql-depth-limit#documentation) to match the name, or a function that returns a boolean.

For example:

> ```js

Reviewer: Again here... I think we could include just the logger config and take out the rest of the irrelevant code
